File: app/channels/conversation_channel.rb
class ConversationChannel < ApplicationCable::Channel
  def subscribed
    conversation = Conversation.find(params[:conversation_id])
    
    # Verify user has access to this conversation
    if current_user&.conversations&.include?(conversation)
      stream_for conversation
    else
      reject
    end
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end
  
  def speak(data)
    conversation = Conversation.find(params[:conversation_id])
    return unless current_user&.conversations&.include?(conversation)
    
    message = conversation.chat_messages.create!(
      sender: current_user,
      content: data['content']
    )
    
    ConversationChannel.broadcast_to(
      conversation,
      {
        message: ApplicationController.render(
          partial: 'conversations/message',
          locals: { message: message, current_user: current_user }
        ),
        sender_id: current_user.id
      }
    )
  end
  
  def start_typing
    conversation = Conversation.find(params[:conversation_id])
    return unless current_user&.conversations&.include?(conversation)
    
    ConversationChannel.broadcast_to(
      conversation,
      {
        typing: true,
        sender_id: current_user.id,
        sender_name: current_user.display_name
      }
    )
  end
  
  def stop_typing
    conversation = Conversation.find(params[:conversation_id])
    return unless current_user&.conversations&.include?(conversation)
    
    ConversationChannel.broadcast_to(
      conversation,
      {
        typing: false,
        sender_id: current_user.id
      }
    )
  end
end

File: app/javascript/channels/conversation_channel.js
import consumer from "./consumer"

// Make App global so we can access it from views
window.App = window.App || {};
window.App.cable = consumer;

// Define the ConversationChannel class for dynamic subscription creation
window.App.ConversationChannel = {
  subscribe: function(conversationId) {
    return consumer.subscriptions.create({
      channel: "ConversationChannel",
      conversation_id: conversationId
    }, {
      received: function(data) {
        if (data.typing !== undefined) {
          // Handle typing indicator
          const typingIndicator = document.getElementById('typing-indicator');
          const typingUser = document.getElementById('typing-user');
          
          if (data.typing && data.sender_id !== window.currentUserId) {
            if (typingUser) typingUser.textContent = data.sender_name;
            if (typingIndicator) typingIndicator.classList.remove('hidden');
          } else if (!data.typing) {
            if (typingIndicator) typingIndicator.classList.add('hidden');
          }
        } else if (data.sender_id !== window.currentUserId) {
          // Handle incoming message
          const messagesContainer = document.getElementById('messages-container');
          if (messagesContainer) {
            messagesContainer.insertAdjacentHTML('beforeend', data.message);
            messagesContainer.scrollTop = messagesContainer.scrollHeight;
          }
        }
      },
      
      speak: function(data) {
        this.perform('speak', data);
        
        // Add message optimistically
        const messagesContainer = document.getElementById('messages-container');
        if (messagesContainer) {
          const messageHtml = `
            <div class="flex justify-end mb-3">
              <div class="max-w-xs lg:max-w-md px-4 py-2 rounded-lg bg-blue-600 text-white rounded-br-sm relative">
                <p class="text-sm whitespace-pre-wrap">${data.content}</p>
                <div class="flex items-center justify-end mt-1 space-x-1">
                  <p class="text-xs text-blue-100">Just now</p>
                  <div class="text-blue-100 text-xs">
                    <i class="fas fa-check"></i>
                  </div>
                </div>
              </div>
            </div>
          `;
          messagesContainer.insertAdjacentHTML('beforeend', messageHtml);
          messagesContainer.scrollTop = messagesContainer.scrollHeight;
        }
      },
      
      startTyping: function() {
        this.perform('start_typing');
      },
      
      stopTyping: function() {
        this.perform('stop_typing');
      }
    });
  }
};
